src/transform/src/dataflow.rs

```diff
@@ -434,76 +448,382 @@ pub fn optimize_dataflow_monotonic(dataflow: &mut DataflowDesc) -> Result<(), Tr
+            }
+        });
 
     mz_repr::explain::trace_plan(dataflow);
 
     Ok(())
 }
+
+#[derive(Debug)]
+struct CollectIndexRequests<'a> {
+    // We were told about these indexes being available.
+    indexes_available: &'a dyn IndexOracle,
+    // We'll be collecting index requests here.
+    index_reqs_by_id: &'a mut BTreeMap<GlobalId, Vec<(Vec<MirScalarExpr>, IndexUsageType)>>,
+    // As we recurse down a MirRelationExpr, we'll need to keep track of the context of the
+    // current node (see docs on `IndexUsageContext` about what context we keep).
+    // Moreover, we need to propagate this context from cte uses to cte definitions.
+    // `context_across_lets` will keep track of the contexts that reached each use of a LocalId
+    // added together.
+    context_across_lets: BTreeMap<LocalId, Vec<IndexUsageContext>>,
+    recursion_guard: RecursionGuard,
+}
+
+impl<'a> CheckedRecursion for CollectIndexRequests<'a> {
+    fn recursion_guard(&self) -> &RecursionGuard {
+        &self.recursion_guard
+    }
+}
+
+impl<'a> CollectIndexRequests<'a> {
+    fn new(
+        indexes_available: &'a dyn IndexOracle,
+        index_reqs_by_id: &'a mut BTreeMap<GlobalId, Vec<(Vec<MirScalarExpr>, IndexUsageType)>>,
+    ) -> CollectIndexRequests<'a> {
+        CollectIndexRequests {
+            indexes_available,
+            index_reqs_by_id,
+            context_across_lets: BTreeMap::new(),
+            recursion_guard: RecursionGuard::with_limit(RECURSION_LIMIT),
+        }
+    }
+
+    pub fn collect_index_reqs(
+        &mut self,
+        expr: &MirRelationExpr,
+    ) -> Result<(), RecursionLimitError> {
+        assert!(self.context_across_lets.is_empty());
+        self.collect_index_reqs_inner(
+            expr,
+            &IndexUsageContext::from_usage_type(IndexUsageType::PlanRoot),
+        )?;
+        assert!(self.context_across_lets.is_empty());
+        Ok(())
+    }
+
+    fn collect_index_reqs_inner(
+        &mut self,
+        expr: &MirRelationExpr,
+        contexts: &Vec<IndexUsageContext>,
+    ) -> Result<(), RecursionLimitError> {
+        // See comment on `IndexUsageContext`.
+        Ok(match expr {
+            MirRelationExpr::Let { id, value, body } => {
+                let shadowed_context = self.context_across_lets.insert(id.clone(), Vec::new());
+                // No shadowing in MIR
+                assert!(shadowed_context.is_none());
+                // We go backwards: Recurse on the body and then the value.
+                self.collect_index_reqs_inner(body, contexts)?;
+                // The above call filled in the entry for `id` in `context_across_lets` (if it
+                // was referenced). Anyhow, at least an empty entry should exist, because we started
+                // above with inserting it.
+                self.collect_index_reqs_inner(
+                    value,
+                    &self.context_across_lets.get(id).unwrap().clone(),
```

Comment:
Sorry, I don't understand how that would allow for getting rid of the `.clone()`.

I guess instead of cloning I could move out from `context_across_lets` with a `remove` (because this info won't be needed again) both in the current code and with a method parameter, but I didn't want to complicate the code, because then I would have to take care that recursive references are not surprised by the entry not being present.